pkg/promtail/targets/gcplog/target.go

```diff
@@ -0,0 +1,165 @@
+package gcplog
+
+import (
+	"context"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/grafana/loki/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/pkg/promtail/targets/target"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/relabel"
+)
+
+var (
+	gcplogEntries = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "gcplog_target_entries_total",
+		Help:      "Help number of successful entries sent to the gcplog target",
+	}, []string{"project"})
+
+	gcplogErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "gcplog_target_parsing_errors_total",
+		Help:      "Total number of parsing errors while receiving gcplog messages",
+	}, []string{"project"})
+)
+
+// GcplogTarget represents the target specific to GCP project.
+// It collects logs from GCP and push it to Loki.
+type GcplogTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrapeconfig.GcplogTargetConfig
+	relabelConfig []*relabel.Config
+	jobName       string
+
+	// lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// pubsub
+	ps   *pubsub.Client
+	msgs chan *pubsub.Message
+}
+
+// NewGcplogTarget returns the new instannce of GcplogTarget for
+// the given `project-id`. It scraps logs from the GCP project
+// and push it Loki via given `api.EntryHandler.`
+// It starts the `run` loop to consume log entries that can be
+// stopped via `target.Stop()`
+func NewGcplogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GcplogTargetConfig,
+) (*GcplogTarget, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ps, err := pubsub.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	target, err := newGcplogTarget(logger, handler, relabel, jobName, config, ps, ctx, cancel)
+	if err != nil {
+		return nil, err
+	}
+
+	go target.run()
+
+	return target, nil
+}
+
+func newGcplogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GcplogTargetConfig,
+	pubsubClient *pubsub.Client,
+	ctx context.Context,
+	cancel func(),
+) (*GcplogTarget, error) {
+
+	pt := &GcplogTarget{
+		logger:        logger,
+		handler:       handler,
+		relabelConfig: relabel,
+		config:        config,
+		jobName:       jobName,
+		ctx:           ctx,
+		cancel:        cancel,
+		ps:            pubsubClient,
+		msgs:          make(chan *pubsub.Message),
+	}
+
+	return pt, nil
+}
+
+func (t *GcplogTarget) run() error {
+	send := t.handler.Chan()
```

Comment:
I think you should defer t.cancel() to stop the pubsub client goroutine. Currently if you break out, this goroutine is not stopped.

Reply:
if you mean this goroutine.
```
	go func() {
		err := sub.Receive(t.ctx, func(ctx context.Context, m *pubsub.Message) {
			t.msgs <- m
		})
		if err != nil {
			// TODO(kavi): Add proper error propagation maybe?
			level.Error(t.logger).Log("error", err)
			gcplogErrors.WithLabelValues(t.config.ProjectID).Inc()
		}
	}()

```

`Stop()` should stop it as well, as we use same `ctx` to `sub.Receive`. Calling Stop() (is the currently only exit point in the run() method.). 

Not sure if that's clear, or Am I missing something?

Reply:
No that's fine I've been mislead by the break below.